elasticdl/python/master/master.py

```diff
@@ -0,0 +1,401 @@
+                if self.task_d.finished():
+                    if self.worker_manager:
+                        self.worker_manager.update_status(
+                            WorkerManagerStatus.FINISHED
+                        )
+                    if self.checkpoint_output_path:
+                        self.master_servicer.save_latest_checkpoint(
+                            self.checkpoint_output_path
+                        )
+                    break
+                time.sleep(30)
+        except KeyboardInterrupt:
+            self.logger.warning("Server stopping")
+
+        self._stop()
+
+    def _stop(self):
+        '''
+        Stop all the components.
+        Make sure that the entire cluster is shut down.
+        '''
+        self.logger.info("Stopping master")
+
+        if self.evaluation_service:
+            self.logger.info("Stopping evaluation service")
+            self.evaluation_service.stop()
+
+        self.logger.info("Stopping RPC server")
+        self.server.stop(0)
+
+        # Keep TensorBoard running when all the tasks are finished
+        if self.tb_service:
+            self.logger.info(
+                "All tasks finished. Keeping TensorBoard service running..."
+            )
+            while True:
+                if self.tb_service.is_active():
+                    time.sleep(10)
+                else:
+                    self.logger.warning(
+                        "Unable to keep TensorBoard running. "
+                        "It has already terminated"
+                    )
+                    break
+        self.logger.info("Master stopped")
+
+    @staticmethod
+    def _get_job_type(args):
+        if all(
+            (
+                args.training_data,
+                args.evaluation_data,
+                args.evaluation_throttle_secs or args.evaluation_steps,
+            )
+        ):
+            job_type = JobType.TRAINING_WITH_EVALUATION
+        elif all(
+            (
+                args.evaluation_data,
+                not args.training_data,
+                not args.prediction_data,
+            )
+        ):
+            job_type = JobType.EVALUATION_ONLY
+        elif all(
+            (
+                args.prediction_data,
+                not args.evaluation_data,
+                not args.training_data,
+            )
+        ):
+            job_type = JobType.PREDICTION_ONLY
+        else:
+            job_type = JobType.TRAINING_ONLY
+
+        return job_type
+
+    def _create_tensorboard_service(self, tensorboard_log_dir, master_ip):
+        tb_service = None
+        if tensorboard_log_dir:
+            self.logger.info(
+                "Create TensorBoard service with log directory %s",
+                tensorboard_log_dir,
+            )
+            # Start TensorBoard CLI
+            tb_service = TensorboardService(tensorboard_log_dir, master_ip)
+
+        return tb_service
+
+    def _create_checkpoint_service(self, args):
```

Comment:
This and other functions below all use `args`. This will make the requirements of args for each function unclear. Perhaps make then more explicit?

Reply:
These sub methods almost use the same parameters with the internal Api call. And for _create_worker_manager, it contains almost all the K-V in args. So at the present, I choose pass args as the private method params.

The sub components will be re-designed in the following development. Will refine the parameters at that time.